runtime/gc_vlhgc/SchedulingDelegate.cpp

```diff
@@ -523,44 +632,188 @@ MM_SchedulingDelegate::calculateEstimatedGlobalBytesToScan() const
+	 * between % of time spent in gc, and pgc pause times. The goal is to minimize % of time spent in gc,
+	 * while staying below the specific gc pause time threshold
+	 */
+
+	uint64_t avgPgcTimeUs = _historicalPartialGCTime * 1000;
+	/* 
+	 * Since _averagePgcInterval measures from start of one PGC to the next, we subtract the avg PGC duration
+	 * to get the avg time between end and start of consecutive PGC's
+	 */
+	uint64_t avgPgcIntervalUs = _averagePgcInterval - avgPgcTimeUs;
+	uintptr_t currentIdealEdenSize = getIdealEdenSizeInBytes(env);
+	uintptr_t currentHeapSize = _regionManager->getRegionSize() * _numberOfHeapRegions;
+	
+	double freeTenureHeadroom = 0.75;
+	
+	/*  
+	 * _estimatedFreeTenure is free space outside of eden and survivor space, plus some additional headroom.
+	 *  We add additional headroom so that we don't ever exhaust that free space
+	 */
+	uintptr_t freeTenure = OMR_MAX((uintptr_t)(_estimatedFreeTenure * freeTenureHeadroom), 1);
+
+	if (0 == _totalGMPWorkTimeUs) {
+		/* We haven't seen a GMP yet, so _estimatedFreeTenure will still be 0, which is not accurate. Use another estimate for free tenure until a GMP happens*/
+		intptr_t freeTenureFromPGCInfo = (intptr_t)currentHeapSize - currentIdealEdenSize - _liveSetBytesAfterPartialCollect - (intptr_t)_averageSurvivorSetRegionCount;
+		freeTenure = freeTenureFromPGCInfo > 0 ? freeTenureFromPGCInfo : 1;
+	}
+
+	Assert_MM_true(freeTenure != 0);
+
+	/* Determine how far we can increase or decrease eden from where eden currently stands. */
+	intptr_t minEdenChange = (intptr_t)currentIdealEdenSize * -1; 
+	intptr_t maxEdenChange = (intptr_t)freeTenure;
+
+	/* How many samples we want to test between minEdenChange and maxEdenChange? */
+	uintptr_t numberOfSamples = 100;
+
+	/* 
+	 * Initially, we suggest the current eden size as the best size - until proven there is a better size.
+	 * The "better" size, will have a better blend of gc overhead (% of time gc is active relative to mutator), 
+	 * and more satisfactory pgc pause time (below target pgc pause is the goal).
+	 */
+	intptr_t recommendedEdenChange = 0;
+	double currentCpuEdenOverhead = predictCpuOverheadForEdenSize(env, currentIdealEdenSize, recommendedEdenChange, freeTenure, avgPgcIntervalUs);
+	double currentEdenHybridOverhead = calculateHybridEdenOverhead(env, _historicalPartialGCTime, currentCpuEdenOverhead);
+	double bestOverheadPrediction = currentEdenHybridOverhead;
+
+	/* How large the hops (in bytes) between samples should be */
+	uintptr_t samplingGranularity = (uintptr_t)(maxEdenChange - minEdenChange) / numberOfSamples;
+
+	/* Try "numberOfSamples" tests on the hybrid overhead curve, to determine which eden change will have best hybrid overhead */
+	for (uintptr_t i = 0; i < numberOfSamples; i++) {
+		/* Start from the right side of the curve */
+		intptr_t edenChange = (intptr_t)maxEdenChange - (samplingGranularity * i);
```

Comment:
I'm missing something here - how do we evaluate for contraction (negative edenChange)!? Looks to me edenChange is always > minEdenChange (what in turn is > 0).

Reply:
`minEdenChange` is simply equal to the negative of current eden size - meaning we are not allowed to try to shrink eden more than the actual size of eden. Consider the following.

5G heap, 1G live, 2G eden.

Eden can grow by at most be 2G (minus a little bit of room for survivor, more like ~1.5 G) -> `maxEdenChange `, and can be shrink by no more than 2G -> `minEdenChange` (which will always be negative).

This means that the change in eden size, is somewhere between -2G and 1.5G

Reply:
> minEdenChange is simply equal to the negative of current eden size
ah, I missed the negative part.